lib/elixir/src/elixir_fn.erl

```diff
@@ -112,6 +112,10 @@ capture_expr(Meta, Expr, S, E, Escaped, Sequential) ->
   case escape(Expr, E, Escaped) of
     {_, []} when not Sequential ->
       invalid_capture(Meta, Expr, E);
+    {EExpr, []} ->
+      Meta2 = lists:keydelete(no_parens, 1, Meta),
```

Comment:
This is not strictly needed for the fix, but otherwise we're expanding to something like ` {:fn, [{:no_parens, true}], [{:->, [{:no_parens, true}], ...` which feels weird and make it harder to test.

Reply:
I think ideally we want to compile this to:

    Function.capture(expr, name, arity)

So something we get something like this back:

```
iex> &(if true, do: NaiveDateTime, else: DateTime).utc_now/0
&NaiveDateTime.utc_now/0
```

Note we already do this for variables:

```
iex> var = NaiveDateTime
iex> &var.utc_now/0
&NaiveDateTime.utc_now/0
```

So we probably need to generalize this branch of the code a bit more (i.e. it should apply whenever the left side of the `.` is an expression)?

Reply:
Nah, ignore me, there is clearly a semantic difference here. Your version executes `foo()` in `&foo().bar/0` when the function is invoked. Mine would execute it always.

I think this is actually a bug. We should not have accepted this format but we clearly do. My suggestion is to deprecate this feature in main.

Reply:
To be clear beyond doubt:

    &expr.fun/arity

Only works if:

  * `expr` is a variable or an atom at compile-time
  * fun is an atom at compile-time
  * arity is an integer at compile-time

If those conditions are not met and it is valid, we should deprecate it.

Reply:
Sounds good, I hesitated to propose something in this vein but was unsure if there would be other legit cases.
As you said assigning to a variable or using fn seems better either way.